### PR TITLE
Init making stampit flexible

### PR DESCRIPTION
**Pull request by koresar**
I didn't update README yet. Will be updated with #71.

@ericelliott, @LegitTalon, @troutowicz please review.

Cheers.

### CODE REVIEW

**Comment by koresar** [comment]
Sorry rushing you up. I intend to finalize v2.0 with few days time. This PR is what holds us down. Does anybody want to take a look? Run tests? Browse the code? I would prefer to run this project collectively if possible.
Cheers

**Comment by troutowicz** [comment]
@koresar Been busy, I can look over it later today.

**Comment by koresar** [comment]
Thanks mate. The main changes to `stampit.js` are around line 131. The corresponding tests are in the `test/init.js`.

File: package.json [hunk 1]
```diff
@@ -32,9 +32,9 @@
   },
   "scripts": {
     "start": "npm run hint && npm run build && npm run uglify",
-    "test": "browserify test/stampit-specs.js | tape-run",
+    "test": "browserify test/stampit-specs.js | node",
```
Comment (troutowicz) [inline package.json]: just make this:

`node test/stampit-specs.js`

Comment (koresar) [inline package.json]: true

**Comment by troutowicz** [comment]
The tests are organized much nicer :+1:

**Comment by koresar** [comment]
I felt like having "npm start" is kind of misleading for a utility module. So I replaced it with more useful "prepublish" hook to make sure none publishes untested code. :)

**Comment by koresar** [comment]
Done.

**Comment by JosephClay** [comment]
Awesome.

I'm running into a jshint error with `test/*`...seems to work passing `test` without the slash+wildcard.

`mixer.js` also has some lint errors.

**Comment by unstoppablecarl** [comment]
Why is `mixer.mixIn` designed to need to use `.apply` couldn't it just use a param instead of `this` and be much more performant? 

https://github.com/ericelliott/stampit/blob/init-making-stampit-flexible/stampit.js#L54

**Comment by JosephClay** [comment]
@unstoppablecarl Would be possible if the arguments were known for [`.mixIn`](https://github.com/ericelliott/stampit/blob/init-making-stampit-flexible/mixer.js#L25)..but I don't think it would be worth duplicating the function for an optimization.

**Comment by unstoppablecarl** [comment]
What about using a faster `.apply` ?

Especially when not using any context.

https://github.com/codemix/fast.js/blob/master/function/applyNoContext.js

**Comment by troutowicz** [comment]
I'm wondering the same @unstoppablecarl. What does the current implementation do differently than:

```js
function addRefs(fixed, states) {
  fixed.refs = fixed.state = mixer.mixIn(fixed.refs || fixed.state, states);
  return fixed.refs;
}
```
And why are we returning `fixed.refs`?

**Comment by troutowicz** [comment]
Actually, is this not doing the same?

```js
function addRefs(fixed, states) {
  mixer.mixIn(fixed.refs || fixed.state, states);
}
```

**Comment by JosephClay** [comment]
@troutowicz No, `objects` is being listed as a parameter, but is not being used. The function is looping through the arguments `Object.assign` style. see: https://github.com/ericelliott/stampit/blob/init-making-stampit-flexible/mixer.js#L25

**Comment by ericelliott** [comment]
> When will you have source.fixed.state and not source.fixed.refs ?
The same for all the other || fallbacks ?

This is support for legacy stamps.

> objects is being listed as a parameter, but is not being used. The function is looping through the arguments Object.assign style. 

This can be fixed after upgrading stampit 2.0 to use ES6 using rest params. Instead of specifying an unused formal param, we can just use `...args` in the function signature, which will give us an array of all the objects.

**Comment by troutowicz** [comment]
@JosephClay If you look at the mixin function, `objects` is never used... `arguments` is actually whats being used in the loop. And as far as I can see, stampit never passes in more than two objects (destination and source). Am I missing something?

**Comment by troutowicz** [comment]
Ah, I see why we are setting `fixed.state` or `fixed.refs` then returning.... its for chaining.

**Comment by ericelliott** [comment]
I wouldn't worry too much about nitpicky implementation details unless you spot something that's really slow. What you should do is:

1. Download it
2. Test & lint
3. Do all the tests pass?
4. Look at all the tests. Do they really cover what they're meant to cover? Do the tests make sense?
5. Look at the implementation. Does it look **correct**? (Note, correct and how you would implement it are two different things).

If you see something and there is clearly a better way, or a learning opportunity for the team, feel free to speak up.

Code reviews shouldn't slow everything down and lead people into bikeshed discussions. They should be a basic code quality protection, just so we all have a bit more peace of mind that the code does what we say it does.

The only thing that matters in software development is that the users love the software. Does it look like it would work and you would love to use it? =)

**Comment by JosephClay** [comment]
@ericelliott Agreed

@troutowicz I think we're both saying the same thing

**Comment by koresar** [comment]
I'll see what I can do about the comments above in 12 hours.

**Comment by koresar** [comment]
> I'm running into a jshint error "cannot open test/*" with test/*...seems to work passing test without the slash+wildcard.

You were probably running old JSHint. No worries. I've fixed that with last commit. `npm run test/index.js`.

**Comment by koresar** [comment]
> I'm wondering the same @unstoppablecarl. What does the current implementation do differently than:

All the proposed implementations above simply break tests. Try it. :)

The tests are very comprehensive, my friends. Very.

**Comment by koresar** [comment]
> What about using a faster .apply ?
I'd prefer later stage optimizations. Premature optimizations are general evil. (I specialize on application performance for few years now.)

Let us think of performance when somebody hit bottleneck and claim it's because of stampit. ;)

Nevertheless, it would be awesome to see a performance optimizing PR after we release v2.0.

In fact, I did analyze performance of stampit. The only bottleneck is the `function factory()`. That's why you can find few `if` statements there. ;)

**Comment by koresar** [comment]
Changes:
* Linting every JS file we wrote.
* Fix linting.
* Rename unusual `mixIn` to `mixin`. Leave alias for compatibility of course.
* Add comments to those weird `||` operators. Hopefully, this will help future code readers.

**Comment by LegitTalon** [comment]
Good stuff. Great discussion, my only concern is documentation but it looks like #83 incapsulates that. :+1:

**Comment by ericelliott** [comment]
Is there an `.assign` alias as well as `.mixin` and `.mixIn`?

**Comment by koresar** [comment]
Yes, there should be one. Thanks for the reminder.

**Comment by ericelliott** [comment]
Are there tests for the aliases? I didn't see them...

**Comment by koresar** [comment]
There are no tests for `stampit.method` aliases as of yet. I wanted to add them though.

**Comment by ericelliott** [comment]
:+1:

**Pull request merged**